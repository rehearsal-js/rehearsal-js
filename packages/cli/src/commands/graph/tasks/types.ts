export interface GraphWorkerOptions {
  rootPath: string;
  srcPath: string;
  ignore: string[];
  serviceMap: Record<string, string>;
  output?: string;
}

type MessageResponse = { type: 'message'; content: string };
type FilesResponse = { type: 'files'; content: string[] };
export type Response = MessageResponse | FilesResponse;
