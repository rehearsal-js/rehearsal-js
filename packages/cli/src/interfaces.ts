export interface GitDescribe {
  tag: string;
  count: number;
  sha: string;
  dirty: boolean;
}
