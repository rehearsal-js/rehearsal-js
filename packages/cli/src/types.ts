import { MigrationStrategy } from '@rehearsal/migration-graph';

import { UserConfig } from './userConfig';
import { State } from './helpers/state';

export type CliCommand = 'upgrade' | 'migrate';
export type Formats = 'sarif' | 'json' | 'sonarqube' | 'md';

export type MigrateCommandOptions = {
  basePath: string;
  entrypoint: string;
  files: string;
  format: Formats[];
  outputPath: string;
  verbose: boolean | undefined;
  userConfig: string | undefined;
  interactive: boolean | undefined;
};

export type MigrateCommandContext = {
  skip: boolean;
  userConfig: UserConfig | undefined;
  strategy: MigrationStrategy;
  sourceFilesWithAbsolutePath: string[];
  sourceFilesWithRelativePath: string[];
  input: unknown;
  targetPackagePath: string;
  state: State;
};

export type UpgradeCommandContext = {
  tsVersion: string;
  latestAvailableBuild: string;
  currentTSVersion: string;
  skip: boolean;
};

export type UpgradeCommandOptions = {
  build: string;
  basePath: string;
  format: Formats[];
  outputPath: string;
  tsVersion: string;
  dryRun: boolean;
};

export type DependencyConfig = {
  dependencies?: string[];
  devDependencies?: string[];
};

export type SetupConfigCommand = {
  command: string;
  args: string[];
};

export type SetupConfig = {
  ts?: SetupConfigCommand;
  lint?: SetupConfigCommand;
};

export type CustomCommandConfig = {
  install?: DependencyConfig;
  setup?: SetupConfig;
};

export type CustomConfig = {
  upgrade?: CustomCommandConfig;
  migrate?: CustomCommandConfig;
};

export type PackageSelection = {
  name: string;
  path: string;
};

export type MenuMap = {
  [key: string]: string;
};

export type MigrationSummary = {
  totalErrorCount: number;
  hintAddedCount: number;
};
export type TSConfig = {
  compilerOptions: {
    strict: boolean;
  };
};

export type ScriptMap = {
  [key: string]: string;
};
