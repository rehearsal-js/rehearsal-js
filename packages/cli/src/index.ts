import { URL } from 'node:url';
import { resolve } from 'node:path';
import assert from 'node:assert';
import fs from 'node:fs/promises';
import { Command } from 'commander';
import { migrateCommand } from './commands/migrate/index.js';
import { upgradeCommand } from './commands/upgrade.js';
import type { PackageJson } from 'type-fest';

const __dirname = new URL('.', import.meta.url).pathname;
const { version } = JSON.parse(
  await fs.readFile(resolve(__dirname, '../package.json'), 'utf-8')
) as PackageJson;

const program = new Command();

assert(version, 'Has a rehearsal version');

program.name('rehearsal').version(version).addCommand(migrateCommand).addCommand(upgradeCommand);

export { program as rehearsal, migrateCommand, upgradeCommand };
