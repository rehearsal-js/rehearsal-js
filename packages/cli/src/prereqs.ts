// ! these should be the bare minimum versions and config required for rehearsal to work with type inference
// ! deps are >= the version specified

import type { PreReqs, ProjectType, PreReqTSConfig } from './types.js';

// node package.json config must be gte 14
const nodeBasePrereq = '14';

// eslint
const eslintBasePrereq = '@typescript-eslint/parser';

// tsconfig
const tsconfigBasePrereq: PreReqTSConfig = {
  compilerOptions: {
    strict: true,
    skipLibCheck: true,
  },
};

// tsconfig ember app/addon
const tsconfigEmberPrereq: PreReqTSConfig = {
  ...tsconfigBasePrereq,
  glint: {
    environment: 'ember-loose',
  },
};

// tsconfig glimmer
const tsconfigGlimmerPrereq: PreReqTSConfig = {
  ...tsconfigBasePrereq,
  glint: {
    environment: 'glimmerx',
  },
};

// pre-req dependencies for rehearsal. doesn't matter if they are deps or devDeps
const depsBasePreReq = {
  typescript: '4.9.0',
  prettier: '2.0.0',
  eslint: '8.0.0',
  'eslint-import-resolver-typescript': '2.5.0',
  'eslint-plugin-import': '2.0.0',
  'eslint-plugin-node': '11.0.0',
  'eslint-plugin-unicorn': '6.0.0',
  '@typescript-eslint/eslint-plugin': '5.0.0',
  '@typescript-eslint/parser': '5.0.0',
};

// pre-req dependencies for ember app/addon
const depsEmberPreReq = {
  ...depsBasePreReq,
  '@glint/core': '1.0.0',
  '@glint/environment-ember-loose': '1.0.0',
  '@glint/environment-ember-template-imports': '1.0.0',
  '@glint/template': '1.0.0',
  'eslint-plugin-ember': '11.0.0',
  'prettier-plugin-ember-template-tag': '0.3.0',
};

// pre-req dependencies for glimmer
const depsGlimmerPreReq = {
  ...depsBasePreReq,
  '@glint/core': '1.0.0',
  '@glint/environment-glimmerx': '1.0.0',
  '@glint/template': '1.0.0',
  'eslint-plugin-ember': '11.0.0',
  '@glimmerx/prettier-plugin-component-templates': '0.6.0',
};

const preReqs: Record<string, PreReqs> = {
  'base-ts': {
    node: nodeBasePrereq,
    eslint: eslintBasePrereq,
    tsconfig: {
      ...tsconfigBasePrereq,
    },
    deps: {
      ...depsBasePreReq,
    },
  },
  ember: {
    node: nodeBasePrereq,
    eslint: eslintBasePrereq,
    tsconfig: {
      ...tsconfigEmberPrereq,
    },
    deps: {
      ...depsEmberPreReq,
    },
  },
  glimmer: {
    node: nodeBasePrereq,
    eslint: eslintBasePrereq,
    tsconfig: {
      ...tsconfigGlimmerPrereq,
    },
    deps: {
      ...depsGlimmerPreReq,
    },
  },
};

export function getPreReqs(type: ProjectType): PreReqs {
  return preReqs[type];
}
