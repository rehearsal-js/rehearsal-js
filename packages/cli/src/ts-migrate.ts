import { migrate, MigrateConfig } from "ts-migrate-server";
import { pluginTSMigrateAutofix } from "@rehearsal-js/plugin-ts-migrate";
import type { Reporter } from "@rehearsal-js/reporter";

export type PluginOptions = { reporter: Reporter };

export async function tsMigrateAutofix(
  srcDir: string,
  reporter: Reporter
): Promise<number> {
  const pluginOptions = { reporter };
  const config = new MigrateConfig().addPlugin(
    pluginTSMigrateAutofix,
    pluginOptions
  );
  const exitCode = await migrate({ rootDir: srcDir, config });

  return exitCode;
}
