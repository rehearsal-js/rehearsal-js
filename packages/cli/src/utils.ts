import { join, resolve } from 'path';
import { readFileSync, writeJSONSync, readJSONSync } from 'fs-extra';
import { compare } from 'compare-versions';
import { parse } from 'json5';
import { valid } from 'semver';
import { type SimpleGit, type SimpleGitOptions, simpleGit } from 'simple-git';
import which from 'which';
import { InvalidArgumentError } from 'commander';

import findup = require('findup-sync');
import execa = require('execa');

import { ScriptMap } from './types';
import type { GitDescribe } from './interfaces';

export const VERSION_PATTERN = /_(\d+\.\d+\.\d+)/;

export const git: SimpleGit = simpleGit({
  baseDir: process.cwd(),
  binary: 'git',
  maxConcurrentProcesses: 6,
} as Partial<SimpleGitOptions>);

// we might need to check if commitlint is being used and lint against it for msgTypes
export async function gitCommit(msg: string, msgType = 'chore(deps-dev)'): Promise<void> {
  await git.commit(`${msgType.trim()}: [REHEARSAL-BOT] ${msg.trim()}`, '--no-verify');
}

// we want a seperate branch & PR for each diagnostic
// eg. await gitCheckoutNewBranch('4.8.0-beta', 'typescript', 'diagnostic-4082')
export async function gitCheckoutNewLocalBranch(
  depSemVersion: string,
  depName = 'typescript',
  addPathID?: string
): Promise<string> {
  const branchName = `rehearsal-bot/${depName}-${depSemVersion}${addPathID ? `/${addPathID}` : ''}`;
  await git.checkout(['-b', branchName]);

  return branchName;
}

export async function gitIsRepoDirty(cwd?: string): Promise<boolean> {
  // if cwd is provided, create a new git instance with the correct cwd
  const gitClient = cwd
    ? simpleGit({
        baseDir: cwd,
        binary: 'git',
        maxConcurrentProcesses: 6,
      } as Partial<SimpleGitOptions>)
    : git;
  if (await gitClient.checkIsRepo()) {
    const status = await gitClient.status();
    return status.isClean() === false;
  }
  return false; // false if it's not a git repo
}

/**
 * Function to introduce a wait
 *
 * @param ms - how many milliseconds to wait
 */
export function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * Convert milliseconds to seconds
 *
 * @param ms - number in milliseconds
 * @returns number in seconds
 */
export function msToSeconds(ms: number): number {
  return Math.round(ms / 1000);
}

export function readJSON<T>(file: string): T | undefined {
  const text = readText(file);
  if (text !== undefined) {
    return parse(text);
  }
}

export function readText(file: string): string | undefined {
  return readFile(file, 'utf8');
}

export function readFile(file: string, encoding: 'utf8'): string | undefined;
export function readFile(file: string, encoding?: undefined): Buffer | undefined;
export function readFile(file: string, encoding?: 'utf8'): string | Buffer | undefined {
  try {
    return readFileSync(file, encoding);
  } catch (e) {
    if ((e as NodeJS.ErrnoException).code === 'ENOENT') {
      return;
    }
    throw e;
  }
}

/**
 * Provide current timestamp
 * @param inSeconds - if true, return seconds, otherwise milliseconds
 * @returns timestamp in seconds
 */
export function timestamp(inSeconds = false): number {
  return inSeconds ? Date.now() / 1000 : Date.now();
}

/**
 * Parse the string output from git
 *
 * @param desc - String output of "git describe --tags --long"
 * @returns GitDescribe object
 */
export function parseLongDescribe(desc: string): GitDescribe {
  const result = /^(.+)-(\d+)-g([a-f0-9]+)(?:-(dirty))?$/.exec(desc);

  if (!result) {
    throw Error(`Unable to parse ${desc} as a long description`);
  }

  const [, tag, count, sha, dirty] = result;

  return {
    tag,
    count: parseInt(count, 10),
    sha,
    dirty: !!dirty,
  };
}

/**
 * Convert instances like "foo-web_10.2.3" and "1.2.3" to "1.2.3"
 *
 * @param versionString - version string to convert into just numbers and periods
 * @returns version string with only numbers and periods
 */
export function normalizeVersionString(versionString: string): string {
  if (VERSION_PATTERN.test(versionString)) {
    const result = VERSION_PATTERN.exec(versionString);
    return result ? result[1] : versionString;
  }
  return versionString;
}

export function determineProjectName(directory = process.cwd()): string | null {
  const packageJSONPath = findup('package.json', {
    cwd: directory,
  });

  if (!packageJSONPath) {
    return null;
  }
  const packageJSON = readJSON<{ name: string }>(packageJSONPath);
  return packageJSON?.name ?? null;
}

export function isYarnManager(): boolean {
  const yarnPath = findup('yarn.lock', {
    cwd: process.cwd(),
  });

  return !!yarnPath;
}

export function isYarnBerryManager(): boolean {
  const lockFilePath = findup('yarn.lock', {
    cwd: process.cwd(),
  });
  const berryConfigPath = findup('.yarnrc.yml', {
    cwd: process.cwd(),
  });

  return !!lockFilePath && !!berryConfigPath;
}

export function isPnpmManager(): boolean {
  const pnpmPath = findup('pnpm-lock.yaml', {
    cwd: process.cwd(),
  });

  return !!pnpmPath;
}

export function getLockfilePath(): string | null {
  const yarnPath = findup('yarn.lock', {
    cwd: process.cwd(),
  });

  if (yarnPath) {
    return yarnPath;
  }

  const pnpmPath = findup('pnpm-lock.yaml', {
    cwd: process.cwd(),
  });

  if (pnpmPath) {
    return pnpmPath;
  }

  const npmPath = findup('package-lock.json', {
    cwd: process.cwd(),
  });

  if (npmPath) {
    return npmPath;
  }

  return null;
}

export function getModuleManager(): 'yarn' | 'npm' | 'pnpm' {
  const isYarn = isYarnManager();
  const isPnpm = isPnpmManager();

  if (isYarn) {
    return 'yarn';
  }

  if (isPnpm) {
    return 'pnpm';
  }

  return 'npm';
}

export function getModuleManagerInstaller(
  manager: 'yarn' | 'npm' | 'pnpm',
  depList: string[],
  isDev: boolean
): { bin: string; args: string[] } {
  const bin = which.sync(manager);

  switch (manager) {
    case 'yarn':
      // since yarn@3 doesn't have --ignore-scripts anymore
      return {
        bin,
        args: isDev
          ? ['add', '-D', ...depList, ...(isYarnBerryManager() ? [] : ['--ignore-scripts'])]
          : ['add', ...depList, ...(isYarnBerryManager() ? [] : ['--ignore-scripts'])],
      };
    case 'pnpm':
      return {
        bin,
        args: isDev ? ['add', '-D', ...depList] : ['add', ...depList],
      };
    case 'npm':
    default:
      return {
        bin,
        args: isDev
          ? ['install', ...depList, '--save-dev', '--ignore-scripts']
          : ['install', ...depList, '--ignore-scripts'],
      };
  }
}

// devDep: dep@version eg. typescript@4.4.4 or typescript or etc
export async function addDep(
  depList: string[],
  isDev: boolean,
  options: execa.Options = {}
): Promise<void> {
  const moduleManager = getModuleManager();
  const binAndArgs = getModuleManagerInstaller(moduleManager, depList, isDev);

  await execa(binAndArgs.bin, [...binAndArgs.args], options);
}

// when executing a command with module manager prefix
// eg. yarn tsc or pnpm tsc or npm tsc
export async function runModuleCommand(args: string[], option: execa.Options = {}): Promise<void> {
  const moduleManager = getModuleManager();
  const binAndArgs = {
    bin: moduleManager,
    args,
  };
  await execa(binAndArgs.bin, [...binAndArgs.args], option);
}

// rather than explicitly setting from node_modules dir we need to handle workspaces use case
// and volta use case
export async function getPathToBinary(
  binaryName: string,
  options: execa.Options = {}
): Promise<string> {
  // pnpm | yarn | npm
  const moduleManager = getModuleManager();
  // /Users/foo/.volta/bin/yarn
  // /usr/local/bin/pnpm
  const moduleManagerBin = which.sync(moduleManager);

  let stdoutMsg;

  try {
    const { stdout } = await execa(moduleManagerBin, ['bin', binaryName], options);
    stdoutMsg = stdout.trim().split(`/${binaryName}`)[0];
  } catch (e) {
    throw new Error(`Unable to find binary path to ${binaryName}`);
  }

  // return the path to the binary
  try {
    return resolve(join(stdoutMsg, binaryName));
  } catch (error) {
    throw new Error(`Unable to find ${binaryName} with ${moduleManagerBin}`);
  }
}

// handles instances where latest is newer than beta and rc etc.
// latestBeta = whichever is latest between latest, beta and rc tags
export async function getLatestTSVersion(tag: string): Promise<string> {
  const module = 'typescript';

  if (tag === 'latestBeta') {
    const compareTags = ['latest', 'beta', 'rc'];
    const versions = await Promise.all(compareTags.map((t) => getLatestModuleVersion(module, t)));
    // compare versions and return the latest
    return versions.reverse().reduce((acc, curr) => {
      if (compare(curr, acc, '>')) {
        return curr;
      }
      return acc;
    }, '0.0.0');
  } else {
    const { stdout } = await execa('npm', ['show', `typescript@${tag}`, 'version']);
    return stdout;
  }
}

export async function getLatestModuleVersion(module: string, tag: string): Promise<string> {
  const { stdout } = await execa('npm', ['show', `${module}@${tag}`, 'version']);

  return stdout;
}

export function isValidSemver(input: string): boolean {
  return !!valid(input);
}

/**
 * Parses the comma separated string `a,b,c` to an array os strings [a, b, c]
 */
export function parseCommaSeparatedList(value: string): string[] {
  return value.split(',');
}

/**
 * Generate tsconfig
 */
export function writeTSConfig(basePath: string, fileList: string[]): void {
  const include = [...fileList.map((f) => f.replace('.js', '.ts'))];
  const config = {
    $schema: 'https://json.schemastore.org/tsconfig',
    compilerOptions: {
      strict: true,
      esModuleInterop: true,
      noUncheckedIndexedAccess: true,
      module: 'es2020',
      moduleResolution: 'node',
      newLine: 'lf',
      target: 'ES2021',
      forceConsistentCasingInFileNames: true,
      noFallthroughCasesInSwitch: true,
      noEmit: true,
    },
    include,
  };

  writeJSONSync(resolve(basePath, 'tsconfig.json'), config, { spaces: 2 });
}

/**
 * Parse the TS Version passed in by the user
 */
export function parseTsVersion(value: string): string {
  if (isValidSemver(value)) {
    return value;
  } else {
    throw new InvalidArgumentError(
      'The tsVersion specified is an invalid string. Please specify a valid version as n.n.n'
    );
  }
}

/**
 * Check if typescript is already installed
 */
export function isTypescriptInDevdep(basePath: string): boolean {
  const packageJSONPath = resolve(basePath, 'package.json');
  const packageJSON = readJSONSync(packageJSONPath);
  return (
    (packageJSON.devDependencies && packageJSON.devDependencies.typescript) ||
    (packageJSON.dependencies && packageJSON.dependencies.typescript)
  );
}

/**
 * Add/Update scripts in package.json
 */
export function addPackageJsonScripts(basePath: string, scriptMap: ScriptMap): void {
  const packageJSONPath = resolve(basePath, 'package.json');
  const packageJSON = readJSONSync(packageJSONPath);
  packageJSON.scripts = { ...packageJSON.scripts, ...scriptMap };
  writeJSONSync(packageJSONPath, packageJSON, { spaces: 2 });
}

/**
 * Run git reset to get all file changes to the previous state
 */
export async function resetFiles(): Promise<void> {
  if (await git.checkIsRepo()) {
    await git.reset(['--hard']);
  }
}
