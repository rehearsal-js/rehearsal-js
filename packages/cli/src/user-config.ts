import { resolve } from 'node:path';
import { addDep } from '@rehearsal/utils';
import { execa } from 'execa';
import { readJSONSync } from 'fs-extra/esm';
import type {
  MigrateCommandConfig,
  UpgradeCommandConfig,
  CliCommand,
  CustomConfig,
} from './types.js';

// Storage and runner for user custom cli config
export class UserConfig {
  public basePath: string;
  public config?: MigrateCommandConfig | UpgradeCommandConfig;

  constructor(basePath: string, command: CliCommand) {
    const config: CustomConfig = readJSONSync(resolve(basePath, 'rehearsal-config.json'));

    this.config = config[command];

    this.basePath = basePath;
  }

  public get hasDependencies(): boolean {
    return (
      !!this.config?.install &&
      (!!this.config?.install?.dependencies || !!this.config?.install?.devDependencies)
    );
  }

  public get dependencies(): string[] {
    if (this.config && this.config.install) {
      const { dependencies } = this.config.install;
      if (dependencies && dependencies.length) {
        return dependencies;
      }
    }
    return [];
  }

  public get devDependencies(): string[] {
    if (this.config && this.config.install) {
      const { devDependencies } = this.config.install;
      if (devDependencies && devDependencies.length) {
        return devDependencies;
      }
    }
    return [];
  }

  public get hasPostInstallHook(): boolean {
    return !!this.config?.postInstall;
  }

  public get hasTsSetup(): boolean {
    return !!this.config?.setup?.ts;
  }

  public get hasPostTsSetupHook(): boolean {
    return !!this.config?.setup?.postTsSetup;
  }

  public get hasLintSetup(): boolean {
    return !!this.config?.setup?.lint;
  }

  public get hasPostLintSetup(): boolean {
    return !!this.config?.setup?.postLintSetup;
  }

  async install(): Promise<void> {
    if (this.config && this.config.install) {
      const { dependencies, devDependencies } = this.config.install;
      if (dependencies && dependencies.length) {
        await addDep(dependencies, false, { cwd: this.basePath });
      }

      if (devDependencies && devDependencies.length) {
        await addDep(devDependencies, true, { cwd: this.basePath });
      }
    }
  }

  async postInstall(): Promise<void> {
    if (this.config && this.config.postInstall) {
      if (Array.isArray(this.config.postInstall)) {
        for (const { command, args } of this.config.postInstall) {
          await execa(command, args, { cwd: this.basePath });
        }
      } else {
        const { command, args } = this.config.postInstall;
        await execa(command, args, { cwd: this.basePath });
      }
    }
  }

  // TODO: how to make setup tasks more generic instead of separated functions?

  // TODO: it doesn't make sense for migrate and upgrade sharing the same setup tasks,
  // What does upgrade command need for custom config?
  async tsSetup(): Promise<void> {
    if (this.config?.setup?.ts) {
      const { ts: tsSetup } = this.config.setup;
      if (Array.isArray(tsSetup)) {
        for (const { command, args } of tsSetup) {
          await execa(command, args, { cwd: this.basePath });
        }
      } else {
        const { command, args } = tsSetup;
        await execa(command, args, { cwd: this.basePath });
      }
    }
  }

  async postTsSetup(): Promise<void> {
    if (this.config?.setup?.postTsSetup) {
      const { postTsSetup } = this.config.setup;
      if (Array.isArray(postTsSetup)) {
        for (const { command, args } of postTsSetup) {
          await execa(command, args, { cwd: this.basePath });
        }
      } else {
        const { command, args } = postTsSetup;
        await execa(command, args, { cwd: this.basePath });
      }
    }
  }

  async lintSetup(): Promise<void> {
    if (this.config?.setup?.lint) {
      const { lint: lintSetup } = this.config.setup;
      if (Array.isArray(lintSetup)) {
        for (const { command, args } of lintSetup) {
          await execa(command, args, { cwd: this.basePath });
        }
      } else {
        const { command, args } = lintSetup;
        await execa(command, args, { cwd: this.basePath });
      }
    }
  }

  async postLintSetup(): Promise<void> {
    if (this.config?.setup?.postLintSetup) {
      const { postLintSetup } = this.config.setup;
      if (Array.isArray(postLintSetup)) {
        for (const { command, args } of postLintSetup) {
          await execa(command, args, { cwd: this.basePath });
        }
      } else {
        const { command, args } = postLintSetup;
        await execa(command, args, { cwd: this.basePath });
      }
    }
  }

  get exclude(): string[] | undefined {
    return (this.config as MigrateCommandConfig)?.exclude;
  }

  get include(): string[] | undefined {
    return (this.config as MigrateCommandConfig)?.include;
  }
}
