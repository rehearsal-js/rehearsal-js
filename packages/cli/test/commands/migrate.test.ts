import { resolve } from 'path';
import { copySync, readdirSync, readFileSync, readJSONSync, writeJSONSync } from 'fs-extra';
import { dirSync, setGracefulCleanup } from 'tmp';
import { beforeAll, beforeEach, describe, expect, test } from 'vitest';

import { CustomConfig } from '../../src/types';
import { runBin } from '../test-helpers';

setGracefulCleanup();

// TODO migrate this to fixturify
const FIXTURE_APP_DIR = resolve(__dirname, '../fixtures/app_for_migrate');
const TEST_SRC_DIR = resolve(FIXTURE_APP_DIR, 'src');

function prepareTmpDir(dir: string): string {
  const srcDir = resolve(TEST_SRC_DIR, dir);
  const { name: targetDir } = dirSync();
  copySync(srcDir, targetDir);
  return targetDir;
}

function createUserConfig(basePath: string, config: CustomConfig): void {
  const configPath = resolve(basePath, 'rehearsal-config.json');
  writeJSONSync(configPath, config);
}

describe('migrate - install dependencies', async () => {
  let basePath = '';

  beforeAll(() => {
    basePath = prepareTmpDir('initialization');
  });

  test('Do install typescript dependency if project dose not have one', async () => {
    const result = await runBin('migrate', [], {
      cwd: basePath,
    });

    expect(result.stdout).toContain('[SUCCESS] Installing dependencies');
  });

  test('Do not install typescript dependency if project already has one', async () => {
    // at this point project already have typescript installed in previous test
    const result = await runBin('migrate', [], {
      cwd: basePath,
    });

    expect(result.stdout).toContain(
      '[SKIPPED] typescript already exists. Skipping installing typescript.'
    );
  });

  test('Install custom dependencies with user config provided', async () => {
    basePath = prepareTmpDir('initialization');
    createUserConfig(basePath, {
      migrate: {
        install: {
          dependencies: ['fs-extra'],
          devDependencies: ['@types/fs-extra'],
        },
      },
    });

    const result = await runBin('migrate', ['-u', 'rehearsal-config.json'], {
      cwd: basePath,
    });

    expect(result.stdout).toContain('Installing custom dependencies');

    const packageJson = readJSONSync(resolve(basePath, 'package.json'));
    expect(packageJson.dependencies).toHaveProperty('fs-extra');
    expect(packageJson.devDependencies).toHaveProperty('@types/fs-extra');
  });
});

describe('migrate - generate tsconfig', async () => {
  let basePath = '';

  beforeAll(() => {
    basePath = prepareTmpDir('initialization');
  });

  test('Create basic tsconfig', async () => {
    const result = await runBin('migrate', [], {
      cwd: basePath,
    });

    expect(result.stdout).toContain('Creating tsconfig');
    expect(readdirSync(basePath)).toContain('tsconfig.json');
  });

  test('Do not create tsconfig if there is an existed one', async () => {
    // tsconfig already created from previous test
    const result = await runBin('migrate', [], {
      cwd: basePath,
    });

    expect(result.stdout).toContain('skipping creating tsconfig.json');
    expect(readdirSync(basePath)).toContain('tsconfig.json');
  });

  test('runBin custom ts config command with user config provided', async () => {
    basePath = prepareTmpDir('initialization');
    createUserConfig(basePath, {
      migrate: {
        setup: {
          ts: { command: 'touch', args: ['custom-ts-config-script'] },
        },
      },
    });

    await runBin('migrate', ['-u', 'rehearsal-config.json'], {
      cwd: basePath,
    });

    expect(readdirSync(basePath)).toContain('custom-ts-config-script');
  });
});

describe('migrate - JS to TS conversion', async () => {
  let basePath = '';

  beforeEach(() => {
    basePath = prepareTmpDir('basic');
  });

  test('able to migrate from default index.js', async () => {
    const result = await runBin('migrate', [], {
      cwd: basePath,
    });

    expect(result.stdout).toContain(`[SUCCESS] Converting JS files to TS`);
    expect(readdirSync(basePath)).toContain('index.ts');
    expect(readdirSync(basePath)).toContain('foo.ts');

    expect(readdirSync(basePath)).not.toContain('index.js');
    expect(readdirSync(basePath)).not.toContain('foo.js');

    const config = readJSONSync(resolve(basePath, 'tsconfig.json'));
    expect(config.include).toContain('index.ts');
    expect(config.include).toContain('foo.ts');
  });

  test('able to migrate from specific entrypoint', async () => {
    console.log(basePath);
    const result = await runBin('migrate', ['--entrypoint', 'depends-on-foo.js'], {
      cwd: basePath,
    });

    expect(result.stdout).toContain(`[SUCCESS] Converting JS files to TS`);
    expect(readdirSync(basePath)).toContain('depends-on-foo.ts');
    expect(readdirSync(basePath)).toContain('foo.ts');

    expect(readdirSync(basePath)).not.toContain('depends-on-foo.js');
    expect(readdirSync(basePath)).not.toContain('foo.js');

    const config = readJSONSync(resolve(basePath, 'tsconfig.json'));
    expect(config.include).toContain('depends-on-foo.ts');
    expect(config.include).toContain('foo.ts');
  });

  test('Print debug messages with verbose', async () => {
    const result = await runBin('migrate', ['--verbose'], {
      cwd: basePath,
    });

    expect(result.stdout).toContain(`\x1B[34mdebug\x1B[39m`);
  });

  test('Generate report with -r flag', async () => {
    await runBin('migrate', ['-r', 'json,md,sarif,foo'], {
      cwd: basePath,
    });

    const reportPath = resolve(basePath, '.rehearsal');

    expect(readdirSync(reportPath)).toContain('report.json');
    expect(readdirSync(reportPath)).toContain('report.md');
    expect(readdirSync(reportPath)).toContain('report.sarif');
    expect(readdirSync(reportPath)).not.toContain('report.foo');
  });
});

describe('migrate - generate eslint config', async () => {
  let basePath = '';

  beforeAll(() => {
    basePath = prepareTmpDir('initialization');
  });

  test('Run custom lint config command with user config provided', async () => {
    basePath = prepareTmpDir('initialization');
    createUserConfig(basePath, {
      migrate: {
        setup: {
          lint: { command: 'touch', args: ['custom-lint-config-script'] },
        },
      },
    });

    const result = await runBin('migrate', ['-u', 'rehearsal-config.json'], {
      cwd: basePath,
    });

    expect(result.stdout).toContain('Creating .eslintrc.js from custom config.');
    expect(readdirSync(basePath)).toContain('custom-lint-config-script');
  });
});

describe('migrate - handle custom basePath', async () => {
  let basePath = '';

  beforeAll(() => {
    basePath = prepareTmpDir('custom_basepath');
  });

  test('Run cli againt specific basePath via -basePath option', async () => {
    const customBasePath = resolve(basePath, 'base');
    const result = await runBin('migrate', ['--basePath', customBasePath], {
      cwd: basePath, // still run cli in basePath
    });

    expect(result.stdout).toContain('[SUCCESS] Installing dependencies');
    expect(result.stdout).toContain('Creating tsconfig');
    expect(readdirSync(customBasePath)).toContain('tsconfig.json');

    expect(result.stdout).toContain(`[SUCCESS] Converting JS files to TS`);
    expect(readdirSync(customBasePath)).toContain('index.ts');
    expect(readdirSync(customBasePath)).not.toContain('index.js');

    const config = readJSONSync(resolve(customBasePath, 'tsconfig.json'));
    expect(config.include).toContain('index.ts');
  });
});
