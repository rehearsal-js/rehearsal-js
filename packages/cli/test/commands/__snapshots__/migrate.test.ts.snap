// Vitest Snapshot v1

exports[`migrate - JS to TS conversion > able to cleanup old JS filse 1`] = `
"import path from 'path';

export function power(foo, bar) {
  return Math.pow(foo, bar);
}"
`;

exports[`migrate - JS to TS conversion > able to migrate multiple JS file from an entrypoint 1`] = `
"export function say(name = 'World') {
  return \`Hello \${name}\`;
}"
`;

exports[`migrate - JS to TS conversion > able to migrate multiple JS file from an entrypoint 2`] = `
"import { say } from './foo';

console.log(say('hello'));
"
`;

exports[`migrate - JS to TS conversion > able to migrate single JS file 1`] = `
"import path from 'path';

export function power(foo, bar) {
  return Math.pow(foo, bar);
}"
`;

exports[`migrate - generate tsconfig > Create strict tsconfig 1`] = `
"{
  \\"$schema\\": \\"http://json.schemastore.org/tsconfig\\",
  \\"compilerOptions\\": {
    \\"baseUrl\\": \\".\\",
    \\"outDir\\": \\"dist\\",
    \\"strict\\": true,
    \\"noUncheckedIndexedAccess\\": true,
    \\"module\\": \\"es2020\\",
    \\"moduleResolution\\": \\"node\\",
    \\"newLine\\": \\"lf\\",
    \\"forceConsistentCasingInFileNames\\": true,
    \\"noFallthroughCasesInSwitch\\": true,
    \\"target\\": \\"es2020\\",
    \\"lib\\": [
      \\"es2020\\"
    ],
    \\"esModuleInterop\\": true,
    \\"declaration\\": true,
    \\"sourceMap\\": true,
    \\"declarationMap\\": true
  },
  \\"include\\": []
}
"
`;
